File: frontend/src/types/swapTrade.ts
/** Core SwapTrade interface matching backend SwapTradeRead schema. */
export interface SwapTrade {
  id: number
  trade_id: string
  portfolio_id: number | null
  counterparty_id: number | null
  portfolio_name: string | null
  counterparty_name: string | null

  // Identifiers
  source_trade_id: string | null
  review_id: string | null

  // Trade details
  ccy_pair: string | null
  direction: string | null      // Buy/Sell | Sell/Buy
  swap_type: string | null
  event_type: string | null
  tenor: string | null
  spread: number | null

  ccy1: string | null
  ccy2: string | null

  // Spot reference
  spot_value_date: string | null
  spot_rate: number | null

  // Near leg
  near_value_date: string | null
  near_tenor: string | null
  near_swap_points: number | null
  near_deal_price: number | null
  near_trade_ccy: string | null
  near_ccy1_amount: number | null
  near_ccy2_amount: number | null
  near_settlement_status: string | null

  // Far leg
  far_value_date: string | null
  far_tenor: string | null
  far_swap_points: number | null
  far_deal_price: number | null
  far_trade_ccy: string | null
  far_ccy1_amount: number | null
  far_ccy2_amount: number | null
  far_settlement_status: string | null

  // Dates
  trade_date: string | null
  trade_time: string | null
  natural_date: string | null

  // Status and meta
  our_trader: string | null
  venue: string | null
  clearing_org: string | null
  clearing_method: string | null
  source: string | null
  comments: string | null

  // Timestamps
  created_at: string | null
  updated_at: string | null
}

export interface SwapTradeListResponse {
  data: SwapTrade[]
  total: number
  page: number
  page_size: number
}

export interface SwapTradeFilterParams {
  page?: number
  page_size?: number
  sort_by?: string
  sort_order?: string
  portfolio_id?: number | null
  counterparty_id?: number | null
  ccy_pair?: string
  direction?: string
  event_type?: string
  trade_date_from?: string
  trade_date_to?: string
  near_value_date_from?: string
  near_value_date_to?: string
  search?: string
}

export interface SwapTradeCreate {
  trade_id: string
  portfolio_id?: number | null
  portfolio_name?: string | null
  counterparty_name?: string | null
  source_trade_id?: string | null
  review_id?: string | null
  ccy_pair?: string | null
  direction?: string | null
  swap_type?: string | null
  event_type?: string | null
  tenor?: string | null
  spread?: number | null
  spot_value_date?: string | null
  spot_rate?: number | null
  near_value_date?: string | null
  near_tenor?: string | null
  near_swap_points?: number | null
  near_deal_price?: number | null
  near_trade_ccy?: string | null
  near_ccy1_amount?: number | null
  near_ccy2_amount?: number | null
  near_settlement_status?: string | null
  far_value_date?: string | null
  far_tenor?: string | null
  far_swap_points?: number | null
  far_deal_price?: number | null
  far_trade_ccy?: string | null
  far_ccy1_amount?: number | null
  far_ccy2_amount?: number | null
  far_settlement_status?: string | null
  trade_date?: string | null
  trade_time?: string | null
  natural_date?: string | null
  our_trader?: string | null
  venue?: string | null
  clearing_org?: string | null
  clearing_method?: string | null
  source?: string | null
  comments?: string | null
}

export interface SwapTradeUpdate {
  trade_id?: string | null
  portfolio_id?: number | null
  portfolio_name?: string | null
  counterparty_name?: string | null
  source_trade_id?: string | null
  review_id?: string | null
  ccy_pair?: string | null
  direction?: string | null
  swap_type?: string | null
  event_type?: string | null
  tenor?: string | null
  spread?: number | null
  spot_value_date?: string | null
  spot_rate?: number | null
  near_value_date?: string | null
  near_tenor?: string | null
  near_swap_points?: number | null
  near_deal_price?: number | null
  near_trade_ccy?: string | null
  near_ccy1_amount?: number | null
  near_ccy2_amount?: number | null
  near_settlement_status?: string | null
  far_value_date?: string | null
  far_tenor?: string | null
  far_swap_points?: number | null
  far_deal_price?: number | null
  far_trade_ccy?: string | null
  far_ccy1_amount?: number | null
  far_ccy2_amount?: number | null
  far_settlement_status?: string | null
  trade_date?: string | null
  trade_time?: string | null
  natural_date?: string | null
  our_trader?: string | null
  venue?: string | null
  clearing_org?: string | null
  clearing_method?: string | null
  source?: string | null
  comments?: string | null
}

File: backend/app/schemas/swap_trade.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict


class SwapTradeBase(BaseModel):
    """Common fields shared across swap trade schemas."""

    trade_id: str | None = None
    portfolio_id: int | None = None
    counterparty_id: int | None = None
    portfolio_name: str | None = None
    counterparty_name: str | None = None

    # Identifiers
    source_trade_id: str | None = None
    review_id: str | None = None

    # Trade details
    ccy_pair: str | None = None
    direction: str | None = None
    swap_type: str | None = None
    event_type: str | None = None
    tenor: str | None = None
    spread: float | None = None

    ccy1: str | None = None
    ccy2: str | None = None

    # Spot reference
    spot_value_date: date | None = None
    spot_rate: float | None = None

    # Near leg
    near_value_date: date | None = None
    near_tenor: str | None = None
    near_swap_points: float | None = None
    near_deal_price: float | None = None
    near_trade_ccy: str | None = None
    near_ccy1_amount: float | None = None
    near_ccy2_amount: float | None = None
    near_settlement_status: str | None = None

    # Far leg
    far_value_date: date | None = None
    far_tenor: str | None = None
    far_swap_points: float | None = None
    far_deal_price: float | None = None
    far_trade_ccy: str | None = None
    far_ccy1_amount: float | None = None
    far_ccy2_amount: float | None = None
    far_settlement_status: str | None = None

    # Dates
    trade_date: date | None = None
    trade_time: str | None = None
    natural_date: date | None = None

    # Status and meta
    our_trader: str | None = None
    venue: str | None = None
    clearing_org: str | None = None
    clearing_method: str | None = None
    source: str | None = None
    comments: str | None = None


class SwapTradeRead(SwapTradeBase):
    """Schema for reading a SwapTrade (API response)."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    trade_id: str
    created_at: datetime | None = None
    updated_at: datetime | None = None


class SwapTradeCreate(SwapTradeBase):
    """Schema for creating a new swap trade manually."""

    trade_id: str


class SwapTradeUpdate(SwapTradeBase):
    """Fields that can be updated on a swap trade."""

    pass


class SwapTradeListResponse(BaseModel):
    """Paginated list response for swap trades."""

    data: list[SwapTradeRead]
    total: int
    page: int
    page_size: int


class SwapTradeFilterParams(BaseModel):
    """Query parameters for filtering swap trades list."""

    page: int = 1
    page_size: int = 50
    sort_by: str = "trade_date"
    sort_order: str = "desc"
    portfolio_id: int | None = None
    counterparty_id: int | None = None
    ccy_pair: str | None = None
    direction: str | None = None
    event_type: str | None = None
    trade_date_from: date | None = None
    trade_date_to: date | None = None
    near_value_date_from: date | None = None
    near_value_date_to: date | None = None
    search: str | None = None
